plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+                                                                        oldDepth
+                                                                        -- for optimization, we pass only a part of the floatdata that are larger than this RHS orig. depth.
+                                                                        (snd $ IM.split oldDepth restDepthTable)
+                                                                      ) rhs
+                                               _ -> error "Something went wrong: no rhs was found for this let in the rhstable."
+                                                       )
+              let (mAnn, mRecurs, newBindings) = foldr1 (<>) $ fmap (\(oAnn,oRec,oBinding,_) -> (oAnn
+                                                                                     -- if the SCC is a single node then use its original 'Recursivity';
+                                                                                     -- otherwise, the SCC is  a group of nodes and we *have to* treat all of them in a 'letrec',
+                                                                                     -- since we don't have any information on how to linearize those.
+                                                                                   , if isSingleton scc then oRec else Rec
+                                                                                     -- lift the binding into a semigroup for accumulation
+                                                                                   , pure oBinding)
+                                                                    ) res
+              pure $ case accTerm of
+                       -- MERGE current let-group with previous let-group iff both groups' recursivity is NonRec
+                       Let accAnn NonRec accBs accIn | mRecurs == NonRec -> Let (mAnn <> accAnn) NonRec (newBindings <> accBs) accIn
+                       -- never merge if the previous let-group is a Rec or the current let-group is Rec,
+                       -- but instead create a nested current let-group under the previous let-group (above)
+                       _ -> Let mAnn mRecurs newBindings accTerm
+
+
+-- | The main transformation function (Term -> Term) that returns a term with all lets floated as outwards as possible (full laziness).
+-- Is comprised of two AST "passes":
+-- 1stpass: to collect the ranks (positions) of all lets
+-- 2ndpass:  to remove all its lets and place them back (float them) in their ranks (positions).
+-- See Note [Float algorithm]
+--
+-- NB: This transformation requires that the PLC.rename compiler-pass has prior been run.
+floatTerm :: forall name tyname uni a
+           . (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique, Monoid a)
+          => Term tyname name uni a -> Term tyname name uni a
+floatTerm pir = p2Term pir
```

Comment:
A rare time when I'd go for the point-free version: `floatTerm = p2Term . p1Term`

Reply:
It won't type check, because `pir` is passed to both the functions, it's not that `p2Term` receives only the result of `p1Term pir`, it also receives `pir`.